plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -199,6 +221,55 @@ newtype ExBudgetMode cost (uni :: GHC.Type -> GHC.Type) fun = ExBudgetMode
     { unExBudgetMode :: forall s. ST s (ExBudgetInfo cost fun s)
     }
 
+{- Note [Cost slippage]
+Tracking the budget usage for every step in the machine adds a lot of overhead. To reduce this,
+we adopt a technique which allows some overshoot of the budget ("slippage"), but only a bounded
+amount.
+
+To do this we:
+- Track all the machine steps of all kinds in an optimized way in a 'WordArray'.
+- Actually "spend" the budget when we've done more than some fixed number of steps, or at the end.
+
+This saves a *lot* of time, at the cost of potentially overshooting the budget by slippage*step_cost,
+which is okay so long as we bound the slippage appropriately.
+
+There are two options for how to bound the slippage:
+1. As a fixed number of steps
+2. As a proportion of the overall budget
+
+Option 2 initially seems much better as a bound: if we run N scripts with an overall budget of B, then
+the potential overrun from 1 is N*slippage, whereas the overrun from 2 is B*slippage. That is, 2
+says we always overrun by a fraction of the total amount of time you were expecting, whereas 1 says
+it depends how many scripts you run... so if I send you a lot of small scripts, I could cause a lot
+of overrun.
+
+However, it turns out (empirically) that we can pick a number for 1 that gives us most of the speedup, but such
+that the maximum overrun is negligible (e.g. much smaller than the "startup cost"). So in the end
+we opted for option 1, which also happens to be simpler to implement.
+-}
+
+{- Note [Structure of the step counter]
+The step counter is kept in a 'WordArray', which is 8 'Word8's packed into a single 'Word64'.
+This happens to suit our purposes exactly, as we want to keep a counter for each kind of step
+that we know about (of which there are 7) and one for the total number.
+
+We keep the counters for each step in the first 7 indices, so we can index them simply by using
+the 'Enum' instance of 'StepKind', and the total counter in the last index.
+
+Why use a 'WordArray'? It optimizes well, since GHC can often do quite a lot of constant-folding
+on the bitwise operations that get emitted. We are restricted to counters of size 'Word8', but this
+is fine since we will reset when we get to 200 steps.
+
+The sharp-eyed reader might notice that the benchmarks in 'word-array' show that 'PrimArray'
+seems to be faster! However, we tried that and it was slower overall, we don't know why.
+-}
+
+type Slippage = Word8
```

Comment:
I googled "slippage" just to make sure that it didn't mean anything unpleasant and I was surprised to discover that it means something in finance, to the extent that most of the hits are about that.  Eg https://en.wikipedia.org/wiki/Slippage_(finance)

Reply:
I was actually vaguely inspired by that. The idea is that you ask us to spend at most X, but we actually spend a little more than X. Which is a bit similar.